src/states/common/base.rs

```diff
@@ -380,32 +375,20 @@ pub trait Base: Display {
         self.send_message(dst, message);
     }
 
-    fn send_message<T: MessageRecipient>(&mut self, dst: &T, message: Message) {
+    fn send_message(&mut self, dst: &ConnectionInfo, message: Message) {
         self.send_message_to_targets(slice::from_ref(dst), 1, message);
     }
 
-    fn send_message_to_targets<T: MessageRecipient>(
+    fn send_message_to_targets(
         &mut self,
-        dst_targets: &[T],
+        conn_infos: &[ConnectionInfo],
         dg_size: usize,
         message: Message,
     ) {
-        let conn_infos: Vec<_> = dst_targets
-            .iter()
-            .filter_map(|dst| dst.resolve().cloned())
-            .collect();
-
         if conn_infos.len() < dg_size {
             warn!(
-                "{} Less than dg_size valid targets! dg_size = {}; targets = {:?}; valid targets = [{:?}]; msg = {:?}",
-                self,
-                dg_size,
-                dst_targets,
-                dst_targets
-                    .iter()
-                    .filter(|dst| dst.resolve().is_some())
-                    .format(", "),
-                message
+                "{} Less than dg_size valid targets! dg_size = {}; targets = {:?}; msg = {:?}",
```

Comment:
Wondering if this warn should be log_or_panic now. Seem that it should never happen.